db/l2db/apiqueries.go

```diff
@@ -104,11 +105,123 @@ func (l2db *L2DB) AddTxAPI(tx *PoolL2TxWrite) error {
 	return nil
 }
 
+// AddAtomicTxsAPI inserts transactions into the pool
+// if minFeeUSD <= total fee in USD <= maxFeeUSD.
+// It's assumed that the given txs conform an atomic group
+func (l2db *L2DB) AddAtomicTxsAPI(txs []PoolL2TxWrite) error {
+	if len(txs) == 0 {
+		return nil
+	}
+	// DB connection handling
+	cancel, err := l2db.apiConnCon.Acquire()
+	defer cancel()
+	if err != nil {
+		return tracerr.Wrap(err)
+	}
+	defer l2db.apiConnCon.Release()
+
+	// Calculate fee in token amount per each used token (don't include tokens with fee 0)
+	feeMap := make(map[common.TokenID]float64)
+	for _, tx := range txs {
+		if _, ok := feeMap[tx.TokenID]; !ok && tx.AmountFloat > 0 {
+			feeMap[tx.TokenID] = tx.Fee.Percentage() * tx.AmountFloat
+		} else {
+			feeMap[tx.TokenID] += tx.Fee.Percentage() * tx.AmountFloat
+		}
+	}
+	tokenIDs := make([]common.TokenID, len(feeMap))
+	pos := 0
+	for id := range feeMap {
+		tokenIDs[pos] = id
+		pos++
+	}
+
+	// Get value in USD for the used tokens (value peer token without decimals)
+	query, args, err := sqlx.In(
+		`SELECT token_id, COALESCE(usd, 0) / (10.0 ^ token.decimals::NUMERIC) AS usd_no_decimals
+		FROM token WHERE token_id IN(?);`,
+		tokenIDs,
+	)
+	if err != nil {
+		return tracerr.Wrap(err)
+	}
+	query = l2db.dbRead.Rebind(query)
+	type tokenUSDValue struct {
+		TokenID     common.TokenID `meddler:"token_id"`
+		USDPerToken float64        `meddler:"usd_no_decimals"`
+	}
+	USDValues := []*tokenUSDValue{}
+	if err := meddler.QueryAll(l2db.dbRead, &USDValues, query, args...); err != nil {
+		return tracerr.Wrap(err)
+	}
+
+	// Calculate average fee per transaction
+	var avgFeeUSD float64 = 0
+	for _, USDValue := range USDValues {
+		avgFeeUSD += feeMap[USDValue.TokenID] * USDValue.USDPerToken
+	}
+	avgFeeUSD = avgFeeUSD / float64(len(txs))
```

Comment:
I think it's better to use the decimal library (https://github.com/shopspring/decimal) for those types of calculations. Float64 is not suitable.

Because float64 (or any binary floating point type, actually) can't represent numbers such as 0.1 exactly.

Consider this code: http://play.golang.org/p/TQBd4yJe6B You might expect that it prints out 10, but it actually prints 9.999999999999831. Over time, these small errors can really add up!

Reply:
Thanks for your suggestion, you're totally right. However I think this is a bit out of the scope of this PR as it would require a relatively big refactor:
- `func (f FeeSelector) Percentage() float64` should return `decimal.Decimal`
- The different types of txs (`Tx`, `PoolL2Tx`, `L2Tx` and `L1Tx`) have a field  `AmountFloat float64` which is used in some places to simplify the code (easier to use than `big` or `decimal` packages). This is also propagated to the DB.

In general we could say that there is no precision safeness across the repo, except when it's truly needed. Therefore I'd recommend to open an issue and tackle the problem from the root. Another related improvement that would emerge is getting rid of the `amount_f NUMERIC` columns in the database. This were used to do numeric calculations directly in SQL when we stored `big.Int` as bytes. However now they're stored as `DECIMAL(78,0)` so there's no need for that anymore.

If you agree, please mark the conversation as resolved and I'll create a detailed issue to do this refactor.

Reply:
A part from that, I think that precision here is not that relevant:
- time between token value updates will be more relevant than precision issues
- This is used only to check if the fee falls between the threshold defined by users using the config file, it's kind of a sanity check. At the end the `txselector` is responsible for making decisions in terms of profitability